File: Frontend/src/types/resume.ts
/**
 * Module: resume.ts
 * Created: 2026-09-03
 * Purpose: TypeScript mirrors of the ATS backend Pydantic schemas.
 */

export interface PersonalInfo {
  name: string;
  email: string;
  phone: string;
  location: string;
  linkedin: string;
  website: string;
}

export interface ExperienceItem {
  company: string;
  title: string;
  location: string;
  start_date: string;
  end_date: string;
  description: string;
}

export interface EducationItem {
  institution: string;
  degree: string;
  start_date: string;
  end_date: string;
  gpa: string;
}

export interface ParsedResumeData {
  personal_info: PersonalInfo;
  summary: string;
  experience: ExperienceItem[];
  education: EducationItem[];
  skills: string[];
  certifications: string[];
  languages: string[];
  projects: Record<string, unknown>[];
}

export interface ResumeOut {
  id: number;
  session_id: string;
  filename: string;
  file_type: string;
  created_at: string;
  updated_at: string;
}

export interface ResumeDetailOut extends ResumeOut {
  parsed_data: ParsedResumeData;
}

export interface GenerateRequest {
  template_id: string;
  format: "pdf" | "docx" | "html";
  parsed_data?: ParsedResumeData;
}

export interface GenerateResponse {
  id: number;
  resume_id: number;
  template_id: string;
  format: string;
  file_path: string;
  created_at: string;
}

export interface TemplateOut {
  id: string;
  name: string;
  style?: string | null;
  layout?: string | null;
  is_custom: boolean;
}

export interface TemplateListOut {
  count: number;
  items: TemplateOut[];
}

File: Backend/app/schemas/resume.py
"""
Module: resume.py
Created: 2026-09-03
Purpose: Pydantic schemas for resume data (request/response) and generation.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class PersonalInfo(BaseModel):
    """Contact and identification details extracted from a CV."""

    name: str = ""
    email: str = ""
    phone: str = ""
    location: str = ""
    linkedin: str = ""
    website: str = ""


class ExperienceItem(BaseModel):
    """A single work experience entry."""

    company: str = ""
    title: str = ""
    location: str = ""
    start_date: str = ""
    end_date: str = ""
    description: str = ""


class EducationItem(BaseModel):
    """A single education entry."""

    institution: str = ""
    degree: str = ""
    start_date: str = ""
    end_date: str = ""
    gpa: str = ""


class ParsedResumeData(BaseModel):
    """The full structured representation of a parsed CV."""

    personal_info: PersonalInfo = Field(default_factory=PersonalInfo)
    summary: str = ""
    experience: list[ExperienceItem] = Field(default_factory=list)
    education: list[EducationItem] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    projects: list[dict] = Field(default_factory=list)


class ResumeCreate(BaseModel):
    """Payload for creating/updating resume structured data."""

    parsed_data: ParsedResumeData


class ResumeUpdate(BaseModel):
    """Payload for editing parsed resume data."""

    parsed_data: ParsedResumeData


class ResumeOut(BaseModel):
    """Read model returned for an uploaded resume."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: str
    filename: str
    file_type: str
    created_at: datetime
    updated_at: datetime


class ResumeDetailOut(ResumeOut):
    """Resume output including its parsed data."""

    parsed_data: dict


class GenerateRequest(BaseModel):
    """Request body for resume generation."""

    template_id: str
    format: str = "pdf"  # pdf | docx | html
    parsed_data: Optional[ParsedResumeData] = None


class GenerateResponse(BaseModel):
    """Response indicating a generated resume file."""

    id: int
    resume_id: int
    template_id: str
    format: str
    file_path: str
    created_at: datetime
